generators/server/templates/src/main/java/package/security/oauth2/AuthorizationHeaderUtil.java.ejs

```diff
@@ -18,46 +18,158 @@
+        OAuth2AccessTokenResponse atr = refreshTokenClient(client);
+        if (atr == null || atr.getAccessToken() == null) {
+            log.info("Failed to refresh token for user");
+            return null;
+        }
+
+        OAuth2RefreshToken refreshToken = atr.getRefreshToken() != null? atr.getRefreshToken(): client.getRefreshToken();
+        OAuth2AuthorizedClient updatedClient = new OAuth2AuthorizedClient(
+            client.getClientRegistration(),
+            client.getPrincipalName(),
+            atr.getAccessToken(),
+            refreshToken
+        );
+
+        clientService.saveAuthorizedClient(updatedClient, oauthToken);
+        return atr.getAccessToken().getTokenValue();
+    }
+
+    private OAuth2AccessTokenResponse refreshTokenClient(OAuth2AuthorizedClient currentClient ) {
+
+        MultiValueMap<String, String> formParameters = new LinkedMultiValueMap<>();
+        formParameters.add(OAuth2ParameterNames.GRANT_TYPE, AuthorizationGrantType.REFRESH_TOKEN.getValue());
+        formParameters.add(OAuth2ParameterNames.REFRESH_TOKEN, currentClient.getRefreshToken().getTokenValue());
+        formParameters.add(OAuth2ParameterNames.CLIENT_ID, currentClient.getClientRegistration().getClientId());
+        RequestEntity requestEntity = RequestEntity
+            .post(URI.create(currentClient.getClientRegistration().getProviderDetails().getTokenUri()))
+            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+            .body(formParameters);
+        try {
+            RestTemplate r = restTemplate(currentClient.getClientRegistration().getClientId(), currentClient.getClientRegistration().getClientSecret());
+            ResponseEntity<OAuthIdpTokenResponseDTO> responseEntity = r.exchange(requestEntity, OAuthIdpTokenResponseDTO.class);
+            return toOAuth2AccessTokenResponse(responseEntity.getBody());
+        } catch (OAuth2AuthorizationException e) {
+            log.error("Unable to refresh token ${e.getError().getErrorCode()}");
+            throw new OAuth2AuthenticationException(e.getError(), e);
+        }
+    }
+
+    private OAuth2AccessTokenResponse toOAuth2AccessTokenResponse(OAuthIdpTokenResponseDTO oAuthIdpResponse) {
+        Map<String, Object> additionalParameters = new HashMap<>();
+        additionalParameters.put("id_token", oAuthIdpResponse.getIdToken());
+        additionalParameters.put("not-before-policy", oAuthIdpResponse.getNotBefore());
+        additionalParameters.put("refresh_expires_in", oAuthIdpResponse.getRefreshExpiresIn());
+        additionalParameters.put("session_state", oAuthIdpResponse.getSessionState());
+        return OAuth2AccessTokenResponse.withToken(oAuthIdpResponse.getAccessToken())
+            .expiresIn(oAuthIdpResponse.getExpiresIn())
+            .refreshToken(oAuthIdpResponse.getRefreshToken())
+            .scopes(Pattern.compile("\\s").splitAsStream(oAuthIdpResponse.getScope()).collect(Collectors.toSet()))
```

Comment:
if we don't use scopes, then, can we remove this?

Reply:
The scope should be at least 'openid' to get a refresh token